### PR TITLE
set scope and build validation in dev when no options provided

### PR DESCRIPTION
**Pull request by wcsanders1**
Resolves issue https://github.com/dotnet/runtime/issues/97091

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs [hunk 1]
```diff
@@ -37,13 +37,15 @@ public partial class HostBuilder : IHostBuilder
         private HostingEnvironment? _hostingEnvironment;
         private IServiceProvider? _appServices;
         private PhysicalFileProvider? _defaultProvider;
+        private readonly bool? _defaultProviderFactoryUsed;
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:40]: Why a nullable here? I don't see where it matters whether this is `false` vs. `null`.

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs [hunk 2]
```diff
@@ -349,6 +351,11 @@ private void InitializeServiceProvider()
                 configureServicesAction(_hostBuilderContext!, services);
             }
 
+            if (_hostBuilderContext!.HostingEnvironment.IsDevelopment() && _defaultProviderFactoryUsed)
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:354]: So the check for `IsDevelopment` is due to this: https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L327

Comment (wcsanders1) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:354]: My understanding is that we don't want those values set to true by default except in development for performance reasons.

**Comment by buyaa-n** [comment]
@benjaminpetit could you take a look please?

**Review by steveharter — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs [hunk 3]
```diff
@@ -37,13 +37,15 @@ public partial class HostBuilder : IHostBuilder
...
         private PhysicalFileProvider? _defaultProvider;
+        private readonly bool _defaultProviderFactoryUsed;
 
         /// <summary>
         /// Initializes a new instance of <see cref="HostBuilder"/>.
         /// </summary>
         public HostBuilder()
         {
             _serviceProviderFactory = new ServiceFactoryAdapter<IServiceCollection>(new DefaultServiceProviderFactory());
+            _defaultProviderFactoryUsed = true;
```
Comment (ericstj) [inline src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:48]: Where is this ever set to false?  Should that be done in `UseServiceProviderFactory`?

**Comment by ericstj** [comment]
I marked this PR as breaking since it introduces new validation by default when in Development environment.

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs (hunk 3 above)
Comment (wcsanders1) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:48]: Yes, it seems to me that this field should be set to false when `UseServiceProviderFactory` is called. This was an oversight.

**Comment by ericstj** [comment]
@steveharter or @wcsanders1 can you please create a breaking change doc for the new defaults for `ServiceProviderOptions` in development?  That will cover the case that we didn't undo here.  It can mention that developers can workaround it using `UseDefaultServiceProvider` if they don't wish to run validation.

**Pull request merged**